### PR TITLE
Ignore files/directories starting with . or _

### LINKED ISSUE #1: Ignore dot and underscore files/directories
Running `tower build` repeatedly in the build directory will yield `_site/_site/_site/...`.

### PR DESCRIPTION
**Pull request by sgrove**
Closes #1

### CODE REVIEW

File: cli/towerSysUtil.ml [hunk 1]
```diff
@@ -41,6 +41,11 @@ let buffer in_file =
   read_file (fun bytes len -> Buffer.add_subbytes buf bytes 0 len) in_file;
   buf
 
+let ignorable_file file_name =
+  (* Currently hard-codes any file starting with . or _ as ignorable by Tower *)
+  let ignorable_pattern = (Str.regexp "^[_\\.]") in
+  Str.string_match ignorable_pattern file_name 0;;
```
Comment (dsheets) [inline cli/towerSysUtil.ml]: This seems like a pretty heavy way to check the first character of a string. I'd rather just use `String.get file_name 0` until we definitely know we need full regexes. Also, this prevents the `str` dep.

File: cli/towerSysUtil.ml [hunk 2]
```diff
@@ -72,7 +77,8 @@ let in_dir path f =
 
 let foldp_paths f p acc dir =
   let dh = Unix.opendir dir in
-  let files = in_dir dir (fun () -> all_files "" [] dh) in
+  let full_files = in_dir dir (fun () -> all_files "" [] dh) in
+  let files = List.filter (fun file_name -> not (ignorable_file file_name)) full_files in
   let () = Unix.closedir dh in
   List.fold_left (fun acc file ->
     if p file dir then f acc file else acc
```
Comment (dsheets) [inline cli/towerSysUtil.ml:78]: I think `ignorable_file` should be used by the caller of `foldp_paths` in the `p` (predicate) parameter which is used here to filter the paths while they are being folded.

File: cli/towerSysUtil.ml (hunk 1 above)
Comment (sgrove) [inline cli/towerSysUtil.ml]: Makes sense, resolved.

File: cli/towerSysUtil.ml (hunk 2 above)
Comment (sgrove) [inline cli/towerSysUtil.ml:78]: Much better, resolved.

**Comment by sgrove** [comment]
If this push looks good, I'll rebase it into a single commit before merging.

**Comment by dsheets** [comment]
This looks pretty good. One thing it won't do that I didn't realize before is ignore paths in subdirectories that begin with `'.'` or `'_'` as `foldp_paths` will construct relative paths from the root. I think to handle that case, we need to change `all_files` to be `filter_files` or some such and pass another (??) predicate into `foldp_paths`. Alternately, `all_files` could return paths as lists of segments and then the single predicate could reject any path with a segment beginning with the hiding characters.

I think I favor this second approach but I haven't really thought a lot about the best way to do this. With segment lists, we'll need to `List.fold_left (/) "" path` in a couple places in `TowerBuild`.

**Comment by dsheets** [comment]
Oh, and a rebase would be most excellent.

**Comment by sgrove** [comment]
Do we want to ignore sub-directories beginning with `.` or `_`? I'll make the change, but not sure of the use case

**Comment by dsheets** [comment]
Subdirectories and files, I think. A lot of tools have subtree environments like git's `.gitignore`. What does Jekyll do?

**Comment by sgrove** [comment]
Looks like Jekyll (likely) also ignores them https://help.github.com/articles/files-that-start-with-an-underscore-are-missing/

Should any of this be accompanied by a test? I see right now the tests are visual rather than unit-test-library-driven

**Comment by dsheets** [comment]
Yes, the inclusion/exclusion of paths should probably be automatically tested. #14 tracks that. The sooner we move away from diff, the better. I've had good experiences with [alcotest](https://github.com/samoht/alcotest).

**Comment by dsheets** [comment]
This looks great! Only comment is `List.exists`/`List.for_all` is cheaper than `List.filter` + match because it can bail out when the predicate is false ASAP. As this code isn't performance critical in the slightest, I'm not too bothered by this.

I'll merge this tomorrow afternoon pending CLA discussion.

**Pull request merged**